### DISCUSSION TITLE
Transfer usdt on Sol

### DISCUSSION THREAD

**Discussion by deeryoona** [trustwallet/wallet-core#3805] [Q&A]
i am an ios developer

example is :
let createAndTransferTokenMessage = SolanaCreateAndTransferToken.with {
$0.recipientMainAddress = "71e8mDsh3PR6gN64zL1HjwuxyKpgRXrPDUJT7XXojsVd"
$0.tokenMintAddress = "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB"
$0.recipientTokenAddress = "EF6L8yJT1SoRoDCkAZfSVmaweqMzfhxZiptKi7Tgj5XY"
$0.senderTokenAddress = "ANVCrmRw7Ww7rTFfMbrjApSPXEEcZpBa6YEiBdf98pAf"
$0.amount = 2900
$0.decimals = 6
}

if accountA is a sender,address is FWu8vwVdmVCb8mxx1pQmhwQXLoFgqDx8LTxsrR84mAvb
accountB is a receiver address is 7i3F63k3jhdKSHC6obRrXkFCzmqnsad9V6QMvRiCcAxR
tokenMintAddress is Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB

there are only 3 parameters,but example is need 4 parameters,I don't know how to set the corresponding parameters

**Comment by satoshiotomakan** [comment]
Hi @deeryoona, please create a discussion, tickets are for bugs/improvements only.

If should clarify that you first need to check if the recipient SOL account has no associated Tether token accounts. You can get them via https://solana.com/docs/rpc/http/gettokenaccountsbyowner. If there is a Token address, you can use the regular `SolanaTokenTransfer`.

Otherwise, you need to generate a new Token address for the recipient:
https://github.com/trustwallet/wallet-core/blob/c0cf03fe36c0b3116b7913bcb199cf3017f966b1/swift/Tests/Blockchains/SolanaTests.swift#L215-L217

The result will be `recipient_token_address` in the `SolanaCreateAndTransferToken` message.

recipient_main_address - recipient SOL account.
token_mint_address - Mint address of the token. Tether token in your case.
recipient_token_address - Token address for the recipient, will be created first.
string sender_token_address - Sender's token address associated with the SOL fee payer.

**Reply by AlexDarsky** [reply]
Hello,I am trying to transfer token to an address that does not have a token account using [ createAndTransferTokenMessage ], then i use [sendTransaction] RPC Method and got the transaction signature.But in the end I couldn’t find the transaction on solscan using the transaction signature.Have you encountered the same problem?Or do you have any ideas or suggestions?

**Reply by satoshiotomakan** [reply]
Hi @AlexDarsky, please try to rebroadcast the transaction multiple times, with the same or new blockhash. That's a common problem of the Solana nodes

**Reply by AlexDarsky** [reply]
OK,I will try it, thank you!

**Comment by ronaldoguedess** [comment]
Thanks @satoshiotomakan , it worked perfectly!
Just one more question: Is it possible to avoid calling the API gettokenaccountsbyowner just to determine which method I should use:
`Solana.TokenTransfer` or `Solana.CreateAndTransferToken`?

In other words, is there an automatic way to handle this?

**Reply by satoshiotomakan** [reply]
Hi @JaimeToca, could you please help to answer?

**Reply by JaimeToca** [reply]
No afaik @ronaldoguedess
